File: sdk/python/src/bf_sdk/resources/layered_configs.py
from __future__ import annotations

from typing import TYPE_CHECKING, Any, cast

from bf_sdk._errors import BFConflict, BFNotFound
from bf_sdk._http import HttpClient
from bf_sdk.models.layered_config import (
    BuildResponse,
    CreateConfigResponse,
    LayeredConfigDetail,
    LayeredConfigListResponse,
    RefreshResponse,
    StoredLayeredConfig,
)
from bf_sdk.models.workload import ResolvedWorkloadRef

if TYPE_CHECKING:
    from bf_sdk.models.workload import WorkloadSummary
    from bf_sdk.runner_config import RunnerConfig


class LayeredConfigs:
    """Advanced low-level layered config management.

    This is kept as a compatibility escape hatch. Most users should prefer
    `BFClient.workloads`, which wraps these control-plane details in a more
    ergonomic workload-first API.
    """

    def __init__(self, http: HttpClient) -> None:
        self._http = http

    def create(self, body: dict[str, Any]) -> CreateConfigResponse:
        data = self._http.post("/api/v1/layered-configs", json_body=body)
        return CreateConfigResponse.model_validate(data)

    def list(self) -> list[StoredLayeredConfig]:
        data = self._http.get("/api/v1/layered-configs")
        payload = dict(data)
        if payload.get("configs") is None:
            payload["configs"] = []
        return LayeredConfigListResponse.model_validate(payload).configs

    def get(self, config_id: str) -> LayeredConfigDetail:
        data = self._http.get(f"/api/v1/layered-configs/{config_id}")
        return LayeredConfigDetail.model_validate(data)

    def delete(self, config_id: str) -> None:
        self._http.delete(f"/api/v1/layered-configs/{config_id}")

    def build(self, config_id: str, *, force: bool = False, clean: bool = False) -> BuildResponse:
        url = f"/api/v1/layered-configs/{config_id}/build"
        params: list[str] = []
        if force:
            params.append("force=true")
        if clean:
            params.append("clean=true")
        if params:
            url += "?" + "&".join(params)
        data = self._http.post(url)
        return BuildResponse.model_validate(data)

    def refresh_layer(self, config_id: str, layer_name: str) -> RefreshResponse:
        data = self._http.post(
            f"/api/v1/layered-configs/{config_id}/layers/{layer_name}/refresh",
        )
        return RefreshResponse.model_validate(data)

    def resolve_workload_ref(
        self,
        config_ref: (
            str
            | CreateConfigResponse
            | StoredLayeredConfig
            | LayeredConfigDetail
            | RunnerConfig
            | WorkloadSummary
        ),
    ) -> ResolvedWorkloadRef:
        direct = self._extract_direct_workload_ref(config_ref)
        if direct.workload_key:
            return direct

        if isinstance(config_ref, LayeredConfigDetail):
            return self._resolve_from_stored_config(config_ref.config)

        if isinstance(config_ref, StoredLayeredConfig):
            return self._resolve_from_stored_config(config_ref)

        ref_value = self._extract_reference_value(config_ref)
        configs = self.list()

        workload_matches = [cfg for cfg in configs if cfg.leaf_workload_key == ref_value]
        if workload_matches:
            return self._resolve_from_stored_config(workload_matches[0])

        config_id_matches = [cfg for cfg in configs if cfg.config_id == ref_value]
        if config_id_matches:
            return self._resolve_from_stored_config(config_id_matches[0])

        display_name_matches = [cfg for cfg in configs if cfg.display_name == ref_value]
        if len(display_name_matches) > 1:
            raise BFConflict(
                f"Multiple layered configs share the display name {ref_value!r}. "
                "Use a config_id or workload key to disambiguate."
            )
        if len(display_name_matches) == 1:
            return self._resolve_from_stored_config(display_name_matches[0])

        raise BFNotFound(
            f"Could not resolve workload {ref_value!r}. "
            "Pass a display name, config_id, create response, or workload key."
        )

    def resolve_workload_key(
        self,
        config_ref: (
            str
            | CreateConfigResponse
            | StoredLayeredConfig
            | LayeredConfigDetail
            | RunnerConfig
            | WorkloadSummary
        ),
    ) -> str:
        """Resolve a user-facing config reference into a control-plane workload key."""
        resolved = self.resolve_workload_ref(config_ref)
        if not resolved.workload_key:
            raise BFNotFound("Resolved workload reference is missing a workload key.")
        return resolved.workload_key

    @staticmethod
    def _extract_reference_value(
        config_ref: (
            str
            | CreateConfigResponse
            | StoredLayeredConfig
            | LayeredConfigDetail
            | RunnerConfig
            | WorkloadSummary
        ),
    ) -> str:
        if isinstance(config_ref, str):
            return config_ref

        if hasattr(config_ref, "display_name"):
            value = cast(Any, config_ref).display_name
            if isinstance(value, str) and value:
                return value

        if hasattr(config_ref, "config_id"):
            value = cast(Any, config_ref).config_id
            if isinstance(value, str) and value:
                return value

        raise BFNotFound("Could not determine how to resolve the requested workload reference.")

    @staticmethod
    def _extract_direct_workload_ref(
        config_ref: (
            str
            | CreateConfigResponse
            | StoredLayeredConfig
            | LayeredConfigDetail
            | RunnerConfig
            | WorkloadSummary
        ),
    ) -> ResolvedWorkloadRef:
        workload_key: str | None = None
        if hasattr(config_ref, "workload_key"):
            value = cast(Any, config_ref).workload_key
            if isinstance(value, str) and value:
                workload_key = value
        if hasattr(config_ref, "leaf_workload_key"):
            value = cast(Any, config_ref).leaf_workload_key
            if isinstance(value, str) and value:
                workload_key = value
        config_id = getattr(config_ref, "config_id", None) if hasattr(config_ref, "config_id") else None
        display_name = getattr(config_ref, "display_name", None) if hasattr(config_ref, "display_name") else None
        return ResolvedWorkloadRef(
            workload_key=workload_key,
            config_id=config_id if isinstance(config_id, str) else None,
            display_name=display_name if isinstance(display_name, str) else None,
        )

    def _resolve_from_stored_config(self, cfg: StoredLayeredConfig) -> ResolvedWorkloadRef:
        if cfg.leaf_workload_key:
            return ResolvedWorkloadRef(
                display_name=cfg.display_name,
                config_id=cfg.config_id,
                workload_key=cfg.leaf_workload_key,
            )
        detail = self.get(cfg.config_id)
        if detail.config.leaf_workload_key:
            return ResolvedWorkloadRef(
                display_name=detail.config.display_name,
                config_id=detail.config.config_id,
                workload_key=detail.config.leaf_workload_key,
            )
        raise BFNotFound(f"Layered config {cfg.config_id!r} does not expose a workload key yet.")

File: sdk/typescript/src/resources/layered-configs.ts
import { CapsuleConflict, CapsuleNotFound } from "../errors";
import type { JsonObject } from "../http";
import { HttpClient } from "../http";
import type {
  BuildResponse,
  CreateConfigResponse,
  LayeredConfigDetail,
  RefreshResponse,
  StoredLayeredConfig,
} from "../models/layered-config";
import { ensureArray } from "../models/common";
import type { ResolvedWorkloadRef, WorkloadSummary } from "../models/workload";
import type { RunnerConfig } from "../runner-config";

export type LayeredConfigReference =
  | string
  | CreateConfigResponse
  | StoredLayeredConfig
  | LayeredConfigDetail
  | RunnerConfig
  | WorkloadSummary;

export class LayeredConfigs {
  public constructor(private readonly http: HttpClient) {}

  public async create(body: Record<string, unknown>): Promise<CreateConfigResponse> {
    return (await this.http.post("/api/v1/layered-configs", {
      jsonBody: body,
    })) as unknown as CreateConfigResponse;
  }

  public async list(): Promise<StoredLayeredConfig[]> {
    const payload = await this.http.get("/api/v1/layered-configs");
    const configs = payload.configs;
    return (Array.isArray(configs) ? configs : []) as StoredLayeredConfig[];
  }

  public async get(configId: string): Promise<LayeredConfigDetail> {
    return (await this.http.get(`/api/v1/layered-configs/${configId}`)) as unknown as LayeredConfigDetail;
  }

  public async delete(configId: string): Promise<void> {
    await this.http.delete(`/api/v1/layered-configs/${configId}`);
  }

  public async build(
    configId: string,
    options: { force?: boolean; clean?: boolean } = {},
  ): Promise<BuildResponse> {
    const params = new URLSearchParams();
    if (options.force) {
      params.set("force", "true");
    }
    if (options.clean) {
      params.set("clean", "true");
    }
    const suffix = params.size > 0 ? `?${params.toString()}` : "";
    return (await this.http.post(`/api/v1/layered-configs/${configId}/build${suffix}`)) as unknown as BuildResponse;
  }

  public async refreshLayer(configId: string, layerName: string): Promise<RefreshResponse> {
    return (await this.http.post(
      `/api/v1/layered-configs/${configId}/layers/${layerName}/refresh`,
    )) as unknown as RefreshResponse;
  }

  public async resolveWorkloadRef(configRef: LayeredConfigReference): Promise<ResolvedWorkloadRef> {
    const direct = extractDirectWorkloadRef(configRef);
    if (direct.workload_key) {
      return direct;
    }

    if (isLayeredConfigDetail(configRef)) {
      return this.resolveFromStoredConfig(configRef.config);
    }

    if (isStoredLayeredConfig(configRef)) {
      return this.resolveFromStoredConfig(configRef);
    }

    const refValue = extractReferenceValue(configRef);
    const configs = await this.list();

    const workloadMatches = configs.filter((config) => config.leaf_workload_key === refValue);
    if (workloadMatches.length > 0) {
      return this.resolveFromStoredConfig(workloadMatches[0]!);
    }

    const configIdMatches = configs.filter((config) => config.config_id === refValue);
    if (configIdMatches.length > 0) {
      return this.resolveFromStoredConfig(configIdMatches[0]!);
    }

    const displayNameMatches = configs.filter((config) => config.display_name === refValue);
    if (displayNameMatches.length > 1) {
      throw new CapsuleConflict(
        `Multiple layered configs share the display name ${JSON.stringify(refValue)}. Use a config_id or workload key to disambiguate.`,
      );
    }
    if (displayNameMatches.length === 1) {
      return this.resolveFromStoredConfig(displayNameMatches[0]!);
    }

    throw new CapsuleNotFound(
      `Could not resolve workload ${JSON.stringify(refValue)}. Pass a display name, config_id, create response, or workload key.`,
    );
  }

  public async resolveWorkloadKey(configRef: LayeredConfigReference): Promise<string> {
    const resolved = await this.resolveWorkloadRef(configRef);
    if (!resolved.workload_key) {
      throw new CapsuleNotFound("Resolved workload reference is missing a workload key.");
    }
    return resolved.workload_key;
  }

  private async resolveFromStoredConfig(config: StoredLayeredConfig): Promise<ResolvedWorkloadRef> {
    if (config.leaf_workload_key) {
      return {
        display_name: config.display_name,
        config_id: config.config_id,
        workload_key: config.leaf_workload_key,
      };
    }

    const detail = await this.get(config.config_id);
    if (detail.config.leaf_workload_key) {
      return {
        display_name: detail.config.display_name,
        config_id: detail.config.config_id,
        workload_key: detail.config.leaf_workload_key,
      };
    }

    throw new CapsuleNotFound(`Layered config ${JSON.stringify(config.config_id)} does not expose a workload key yet.`);
  }
}

function extractReferenceValue(configRef: LayeredConfigReference): string {
  if (typeof configRef === "string") {
    return configRef;
  }

  const json = configRef as unknown as JsonObject;
  const configId = json.config_id;
  if (typeof configId === "string" && configId.length > 0) {
    return configId;
  }

  const displayName = json.display_name;
  if (typeof displayName === "string" && displayName.length > 0) {
    return displayName;
  }

  throw new CapsuleNotFound("Could not determine how to resolve the requested workload reference.");
}

function extractDirectWorkloadRef(configRef: LayeredConfigReference): ResolvedWorkloadRef {
  if (typeof configRef === "string") {
    return {};
  }

  const json = configRef as unknown as JsonObject;
  const workloadKey = typeof json.workload_key === "string"
    ? json.workload_key
    : typeof json.leaf_workload_key === "string"
      ? json.leaf_workload_key
      : undefined;

  return {
    workload_key: workloadKey,
    config_id: typeof json.config_id === "string" ? json.config_id : undefined,
    display_name: typeof json.display_name === "string" ? json.display_name : undefined,
  };
}

function isStoredLayeredConfig(value: LayeredConfigReference): value is StoredLayeredConfig {
  return typeof value === "object" && value !== null && "config_id" in value;
}

function isLayeredConfigDetail(value: LayeredConfigReference): value is LayeredConfigDetail {
  return typeof value === "object" && value !== null && "config" in value;
}
